README.md
# digdag-operator-pg_lock
[![Jitpack](https://jitpack.io/v/pro.civitaspo/digdag-operator-pg_lock.svg)](https://jitpack.io/#pro.civitaspo/digdag-operator-pg_lock) ![Test CI Status Badge](https://github.com/civitaspo/digdag-operator-pg_lock/workflows/Test%20CI/badge.svg) [![Digdag](https://img.shields.io/badge/digdag-v0.9.39-brightgreen.svg)](https://github.com/treasure-data/digdag/releases/tag/v0.9.39)

A digdag plugin to run digdag tasks with locks by PostgreSQL.

# Overview

- Plugin type: operator

# Usage

```yaml
_export:
  plugin:
    repositories:
      - https://jitpack.io
    dependencies:
      - pro.civitaspo:digdag-operator-pg_lock:0.0.3

+lock-with:
  # Wait during 5m until getting the named lock if another task locks.
  # This named lock can be used beyond the workflow and the project,
  # that means if another task in another project locks by the name, this task waits until the lock is released.
  pg_lock>: lock-1
  wait_timeout: 5m
  _do:
    echo>: hello world

+control-parallelism:
  # Tasks can get 5 locks.
  # When parallel tasks run with this feature, you can control the parallelism.
  # The below example means the tasks run in 5 parallels.
  loop>: 100
  _parallel: true
  _do:
    pg_lock>: lock-2  
    limit: 5
    _do:
      echo>: hello world
```

See [example](./example).

# Configuration

## Remarks

- type `DurationParam` is strings matched `\s*(?:(?<days>\d+)\s*d)?\s*(?:(?<hours>\d+)\s*h)?\s*(?:(?<minutes>\d+)\s*m)?\s*(?:(?<seconds>\d+)\s*
s)?\s*`.
  - The strings is used as `java.time.Duration`.
  
## System configurations

- **pg_lock.host**: The host name of PostgreSQL. (string, required)
- **pg_lock.port**: The port of PostgreSQL. (integer, default: `5432`)
- **pg_lock.database**: The database name of PostgreSQL. (string, required)
- **pg_lock.schemata**: The schema (or several schemata separated by commas) to be set in the search-path. (string, default: `"public"`)
- **pg_lock.user**: The user name of PostgreSQL. (string, required)
- **pg_lock.password**: The password of PostgreSQL. (string, optional)
- **pg_lock.login_timeout**: The timeout duration to wait for establishment of a PostgreSQL database connection. (`DurationParam`, default: `30s`)
- **pg_lock.socket_timeout**: The timeout duration for socket read operations. If reading from the server takes longer than this value, the connection is closed. This can be used as both a brute force global query timeout and a method of detecting network problems. `0s` means that it is disabled. (`DurationParam`, default: `30m`)
- **pg_lock.ssl**: (boolean, default: `false`)
- **pg_lock.connection_timeout**: The timeout duration that a client will wait for a connection from the pool. If this time is exceeded without a connection becoming available, a `SQLException` will be thrown. (`DurationParam`, default: `30s`)
- **pg_lock.idle_timeout**: The timeout duration that a connection is allowed to sit idle in the pool. `0s` means that idle connections are never removed from the pool. (`DurationParam`, default: `10m`)
- **pg_lock.validation_timeout**: The timeout duration that the pool will wait for a connection to be validated as alive. (`DurationParam`, default: `5s`)
- **pg_lock.max_pool_size**: The connection pool size that is allowed to reach, including both idle and in-use connections. (integer, default: `5`)
- **pg_lock.min_idle_size**: The property controls the minimum number of idle connections in the pool. (integer, default: the same value as **pg_lock.max_pool_size**)
- **pg_lock.max_life_time**: This property controls the maximum lifetime of a connection in the pool. When a connection reaches this timeout, even if recently used, it will be retired from the pool. An in-use connection will never be retired, only when it is idle will it be removed. (`DurationParam`, default: `30m`)
- **pg_lock.leak_detection_threshold**: The threshold that a connection can be out of the pool before a message is logged indicating a possible connection leak. `0s` means that it is disabled. (`DurationParam`, default: `0s`)
- **pg_lock.poll_interval**: The polling interval to wait for getting the named lock. (`DurationParam`, default: `5s`)
- **pg_lock.min_poll_interval**: The minimum polling interval to wait for getting the named lock. (`DurationParam`, default: `5s`)
- **pg_lock.max_poll_interval**: The maximum polling interval to wait for getting the named lock. (`DurationParam`, default: `5m`)
- **pg_lock.schema_migration**: Whether do schema migration or not. (boolean, default: `true`)
- **pg_loch.schema_migration_history_table**: The table name to write schema migration history. (string, default: `"pg_lock_schema_migrations"`)
- **pg_lock.hash_seed_for_advisory_lock**: The seed to hash strings with "MurmurHash 3" algorithm for `pg_try_advisory_lock`. (integer, default: `-137723950` (the same as `scala.util.hashing.MurmurHash3.stringSeed`))
- **pg_lock.digdag.host**: The digdag server host name. If this option is specified, this operator can check the owner attempt status, and then can release locks that a finished task is the owner of. (string, optional)
- **pg_lock.digdag.port**: The digdag server port. (integer, default: `65432`)
- **pg_lock.digdag.disable_cert_validation**: Disable the cert validation or not. (boolean, default: `false`)
- **pg_lock.digdag.ssl**: Use ssl or not. (boolean, default: `true`)
- **pg_lock.digdag.headers.***: The headers when connecting to the digdag server. You can set multiple values like `pg_lock.digdag.headers.KEY=VALUE`. (string, default: `{}`) 
- **pg_lock.digdag.proxy_schema**: The digdag proxy schema. (string, optional)
- **pg_lock.digdag.proxy_host**: The digdag proxy host name. (string, optional) 
- **pg_lock.digdag.proxy_port**: The digdag proxy port. (integer, optional) 


## Operator configurations

- **pg_lock>**: The name of lock. This name is used across workflows and projects. (string, required)
- **wait_timeout**: The timeout to wait for getting the named lock. (`DurationParam`, default: `15m`)
- **expire_in**: The duration that the named lock expires in. (`DurationParam`, default: `"1h"`)
- **limit**: The limit count of the named locks within the namespace. If the different value is defined in another task, throw `ConfigException`. (integer, default: `1`) 
- **namespace**: The namespace that the named lock can be unique. The valid values are `"global"`, `"site"`, `"project"`, `"workflow"`, `"session"`, and`"attempt"`. (string, default: `"site"`)
- **unlock_finished_attempt_locks**: Unlock the other locks having the same namespace and name as you want to lock if the owner attempts of the locks are finished. (boolean, default: `true`)
- **_do**: The definition of subtasks with the named lock. (config, required) 


# Warnings

* When you cancel the tasks inside `_do` of `pg_lock>`, you must wait for the release of the lock until the expiration. So, I recommend you to set the `expire_in` duration as short as possible.
    * This problem is resolved by [treasure-data/digdag \[feature request\] `_canceled` parameter for some cleanup tasks #1226](https://github.com/treasure-data/digdag/issues/1226).
    * If `pg_lock.digdag.*` propertise are set correctly, this problem is no longer a big problem because this operator can check the owner attempt status, and then can release locks that a finished task is the owner of.
    * You can unlock the lock forcibly by using the internal operator `pg_unlock`. You need to find the lock ID from the log. This operator is used internally, so be careful when using it as changes that are not backward compatible may be made. See the [PgUnlockOperator implementation](./src/main/scala/pro/civitaspo/digdag/plugin/pg_lock/unlock/PgUnlockOperator.scala).
      ```yaml
      +unlock-forcibly:
        pg_unlock>: ${the lock ID}
        force: true
      ```
* This plugin close connections in `PgLockPgConnectionPooler#finalize()`, so the number of connections does not necessarily exceed value of **pg_lock.max_pool_size**.
    * This problem is resolved by [treasure-data/digdag \[feature request\] Lifecycle management plugin interface OR Global resource management plugin interface #1238](https://github.com/treasure-data/digdag/issues/1238)

# Development

## Run an Example

### 1) build

```sh
./gradlew publish
```

Artifacts are build on local repos: `./build/repo`.

### 2) run a PostgreSQL

```sh
./example/run-pg.sh
```

PostgreSQL Runs foreground.

### 3) run examples

```sh
./example/run.sh
```

## Run Tests

### 1) run a PostgreSQL

```sh
./example/run-pg.sh
```

PostgreSQL Runs foreground.

### 2) run tests

```sh
./gradlew test
```

# ChangeLog

[CHANGELOG.md](./CHANGELOG.md)

# License

[Apache License 2.0](./LICENSE.txt)

# Author

@civitaspo



CHANGELOG.md
0.0.3 (2019-11-29)
==================

* [Fix] Fix the way to get `pg_lock.digdag.headers.*` values.

0.0.2 (2019-11-27)
==================

* [Enhancement] Add `unlock_finished_attempt_locks` option.

0.0.1 (2019-11-25)
==================

* [Enhancement] Release locks if owner_attempt is finished.
* [Enhancement] Use Github Actions instead of Circle CI.

0.0.1.pre3 (2019-10-14)
=======================

* [Fix] Fix the connection flood.

0.0.1.pre2 (2019-10-07)
=======================

* [Fix] Fix `org.flywaydb.core.api.FlywayException: Found non-empty schema(s) "public" without schema history table! Use baseline() or set baselineOnMigrate to true to initialize the schema history table.` when using this operator in digdag server mode. [#2](https://github.com/civitaspo/digdag-operator-pg_lock/pull/2)
* [Fix] Fix a namespace bug: `Unsupported namespace: session (config)` [#3](https://github.com/civitaspo/digdag-operator-pg_lock/pull/3)

0.0.1.pre1 (2019-10-05)
=======================

* First Release


src/main/scala/pro/civitaspo/digdag/plugin/pg_lock/lock/PgLockOperator.scala
package pro.civitaspo.digdag.plugin.pg_lock.lock


import java.util.UUID

import com.typesafe.scalalogging.LazyLogging
import io.digdag.client.config.{Config, ConfigException, ConfigFactory}
import io.digdag.client.DigdagClient
import io.digdag.client.api.Id
import io.digdag.spi.{OperatorContext, TaskResult}
import io.digdag.util.{BaseOperator, DurationParam}
import pro.civitaspo.digdag.plugin.pg_lock.pg.{PgLockPgClient, PgLockPgDao}

import scala.jdk.CollectionConverters._
import scala.util.chaining._
import scala.util.hashing.MurmurHash3


class PgLockOperator(context: OperatorContext,
                     systemConfig: Config,
                     pgClient: PgLockPgClient)
    extends BaseOperator(context)
        with LazyLogging
{
    override def runTask(): TaskResult =
    {
        try OperatorRunner().run()
        finally pgClient.close()
    }

    private case class OperatorRunner()
    {
        protected def params: Config =
        {
            request.getConfig
        }

        protected def parseNamespace(ns: String): PgLockNamespace =
        {
            PgLockNamespace.Factory(request).create(ns)
        }

        protected val name: String = params
            .get("_command", classOf[String])
        protected val waitTimeout: DurationParam = params
            .get("wait_timeout", classOf[DurationParam], DurationParam.parse("15m"))
        protected val expireIn: DurationParam = params
            .get("expire_in", classOf[DurationParam], DurationParam.parse("1h"))
        protected val limit: Int = params
            .get("limit", classOf[Int], 1)
        protected val namespace: PgLockNamespace = params
            .get("namespace", classOf[String], "site")
            .pipe(parseNamespace)
        protected val unlockFinishedAttemptLocks: Boolean = params
            .get("unlock_finished_attempt_locks", classOf[Boolean], true)
        protected val doConfig: Config = params
            .get("_do", classOf[Config])

        protected val hashSeedForAdvisoryLock: Int = systemConfig
            .get(s"pg_lock.hash_seed_for_advisory_lock", classOf[Int], -137723950)
        protected val pollInterval: DurationParam = systemConfig
            .get(s"pg_lock.poll_interval", classOf[DurationParam], DurationParam.parse("5s"))
        protected val minPollInterval: DurationParam = systemConfig
            .get(s"pg_lock.min_poll_interval", classOf[DurationParam], DurationParam.parse("5s"))
        protected val maxPollInterval: DurationParam = systemConfig
            .get(s"pg_lock.max_poll_interval", classOf[DurationParam], DurationParam.parse("5m"))
        protected val digdagHost: Option[String] = Option(systemConfig.getOptional("pg_lock.digdag.host", classOf[String]).orNull())
        protected val digdagPort: Int = systemConfig.get("pg_lock.digdag.port", classOf[Int], 65432)
        protected val digdagDisableCertValidation: Boolean = systemConfig.get("pg_lock.digdag.disable_cert_validation", classOf[Boolean], false)
        protected val digdagSsl: Boolean = systemConfig.get("pg_lock.digdag.ssl", classOf[Boolean], true)
        protected val digdagHeaders: Map[String, String] = systemConfig
            .getKeys
            .asScala
            .filter(_.startsWith("pg_lock.digdag.headers."))
            .map(k => k.replace("pg_lock.digdag.headers.", "") -> systemConfig.get(k, classOf[String]))
            .toMap
        protected val digdagProxySchema: Option[String] = Option(systemConfig.getOptional("pg_lock.digdag.proxy_schema", classOf[String]).orNull())
        protected val digdagProxyHost: Option[String] = Option(systemConfig.getOptional("pg_lock.digdag.proxy_host", classOf[String]).orNull())
        protected val digdagProxyPort: Option[Int] = Option(systemConfig.getOptional("pg_lock.digdag.proxy_port", classOf[Int]).orNull())


        protected val cf: ConfigFactory = request.getConfig.getFactory
        protected val pollingWaiter: PgLockPollingWaiter = PgLockPollingWaiter(
            lastStateParams = request.getLastStateParams,
            cf = cf,
            waitTimeout = waitTimeout,
            pollInterval = pollInterval,
            minPollInterval = minPollInterval,
            maxPollInterval = maxPollInterval
            )

        def run(): TaskResult =
        {
            pgClient.transaction { dao =>
                withAdvisoryLock(dao) {
                    releaseExpiredLocks(dao)
                    releaseFinishedAttemptLocks(dao)
                    validateLockLimit(dao)
                    waitUntilAffordingLockableCount(dao)

                    val lockId: UUID = UUID.randomUUID()
                    doLock(dao, lockId)
                    pgClient.commit()
                    logger.info(s"Successfully get the lock (id: $lockId, namespace_type: ${namespace.getType}," +
                                    s" namespace_value: ${namespace.getValue}, owner_attempt_id: ${namespace.getOwnerAttemptId}," +
                                    s" expire_in: ${expireIn.toString}, limit: $limit)")

                    buildTaskResult(lockId)
                }
            }
        }

        protected def hash(str: String): Int =
        {
            MurmurHash3.stringHash(str, hashSeedForAdvisoryLock)
        }

        protected def withAdvisoryLock[A](dao: PgLockPgDao)
                                         (f: => A): A =
        {
            val k1: Int = hash(s"${namespace.getType}:${namespace.getValue}")
            val k2: Int = hash(name)

            pollingWaiter.pollingWait("advisoryLock", dao.pgTryAdvisoryLock(k1, k2))

            try f
            finally {
                try {
                    if (!dao.pgAdvisoryUnlock(k1, k2)) {
                        logger.warn(s"'pg_advisory_unlock($k1, $k2)' is failed.")
                    }
                }
                catch {
                    case ex: Throwable =>
                        logger.warn(
                            s"'pg_advisory_unlock($k1, $k2)' is failed because of ${ex.getMessage}",
                            ex
                            )
                }
            }
        }

        protected def releaseExpiredLocks(dao: PgLockPgDao): Unit =
        {
            dao.deleteExpiredDigdagPgLocks(
                namespaceType = namespace.getType,
                namespaceValue = namespace.getValue,
                name = name
                )
        }

        protected def releaseFinishedAttemptLocks(dao: PgLockPgDao): Unit =
        {
            if (!unlockFinishedAttemptLocks) {
                logger.info("Skip to release the other locks that other attempts are the owner of because unlock_finished_attempt_locks=false.")
                return
            }
            digdagHost.foreach { host =>
                val digdagClient = DigdagClient.builder()
                    .host(host)
                    .port(digdagPort)
                    .ssl(digdagSsl)
                    .headers(digdagHeaders.asJava)
                    .disableCertValidation(digdagDisableCertValidation)
                    .tap { builder =>
                        digdagProxySchema.foreach(builder.proxyScheme)
                        digdagProxyHost.foreach(builder.proxyHost)
                        digdagProxyPort.foreach(builder.proxyPort)
                    }
                    .build()
                try {
                    dao.selectDistinctOwnerAttemptsDigdagPgLocks(
                        namespaceType = namespace.getType,
                        namespaceValue = namespace.getValue,
                        name = name
                        )
                        .asScala
                        .foreach { attemptId =>
                            val attempt = digdagClient.getSessionAttempt(Id.of(attemptId.toString))
                            if (attempt.getDone) {
                                logger.info(s"Release the lock because the attempt(id:${attempt.getId},project:${attempt.getProject.getName}," +
                                                s"workflow:${attempt.getWorkflow.getName},session:${attempt.getSessionUuid.toString}) is already finished.")
                                dao.deleteOwnedDigdagPgLocks(ownerAttemptId = attemptId)
                            }
                        }
                }
                catch {
                    case ex: Throwable =>
                        logger.warn(s"Failed to release the finished attempt locks because of ${ex.getMessage}", ex)
                }
                finally {
                    digdagClient.close()
                }
            }
        }

        protected def validateLockLimit(dao: PgLockPgDao): Unit =
        {
            dao.selectDistinctLimitCountDigdagPgLocks(
                namespaceType = namespace.getType,
                namespaceValue = namespace.getValue,
                name = name
                )
                .asScala
                .foreach { i =>
                    if (i != limit) {
                        throw new ConfigException(
                            s"Conflict current config: limit=$limit" +
                                s" because another workflow defines limit=$i."
                            )
                    }
                }
        }

        protected def waitUntilAffordingLockableCount(dao: PgLockPgDao): Unit =
        {
            pollingWaiter.pollingWait("affordLockableCount", {
                val cnt: Int = dao.selectCountDigdagPgLocks(
                    namespaceType = namespace.getType,
                    namespaceValue = namespace.getValue,
                    name = name
                    )
                ((cnt + 1) <= limit).tap { isLockable =>
                    if (!isLockable) {
                        logger.info(
                            s"Wait because current lock count=$cnt reaches the limit=$limit." +
                                s" (namespace_type: ${namespace.getType}," +
                                s" namespace_value: ${namespace.getValue}, name: $name)"
                            )
                    }
                }
            })
        }

        protected def doLock(dao: PgLockPgDao,
                             lockId: UUID): Unit =
        {
            dao.insertDigdagPgLock(
                id = lockId,
                namespaceType = namespace.getType,
                namespaceValue = namespace.getValue,
                ownerAttemptId = namespace.getOwnerAttemptId,
                name = name,
                limitCount = limit,
                expireInSeconds = expireIn.getDuration.getSeconds
                )
        }

        protected def buildTaskResult(lockId: UUID): TaskResult =
        {
            TaskResult.defaultBuilder(request)
                .subtaskConfig(
                    cf.create()
                        .setNested("_error", buildSubtaskPgUnlock(lockId))
                        .setNested("+do", doConfig)
                        .setNested("+finally", buildSubtaskPgUnlock(lockId))
                    )
                .build()
        }

        protected def buildSubtaskPgUnlock(lockId: UUID): Config =
        {
            cf.create().tap { subtask =>
                subtask
                    .set("_type", "pg_unlock")
                    .set("_command", lockId)
                    .getNestedOrSetEmpty("_retry")
                    // TODO: configurable?
                    .set("limit", 30)
                    .set("interval", 1)
                    .set("max_interval", 10)
                    .set("interval_type", "exponential")
            }
        }
    }
}
